### PR TITLE
unbox closures

### PR DESCRIPTION
**Pull request by apasel422**
This will be necessary once https://github.com/rust-lang/rust/pull/20578 lands.

### CODE REVIEW

File: src/trie/map.rs [hunk 1]
```diff
@@ -187,8 +187,8 @@ impl<T> TrieMap<T> {
     /// assert_eq!(vec, vec!["c", "b"]);
     /// ```
     #[inline]
-    pub fn each_reverse<'a>(&'a self, f: |&uint, &'a T| -> bool) -> bool {
-        self.root.each_reverse(f)
+    pub fn each_reverse<F: FnMut(&uint, &T) -> bool>(&self, mut f: F) -> bool {
```
Comment (Gankro) [inline src/trie/map.rs]: I've been leaning towards a convention with Fn's all in Where clauses, personally.

**Comment by apasel422** [comment]
Converted to where clauses.

File: src/trie/map.rs [hunk 2]
```diff
@@ -187,8 +187,8 @@ impl<T> TrieMap<T> {
     /// assert_eq!(vec, vec!["c", "b"]);
     /// ```
     #[inline]
-    pub fn each_reverse<'a>(&'a self, f: |&uint, &'a T| -> bool) -> bool {
-        self.root.each_reverse(f)
+    pub fn each_reverse<F>(&self, mut f: F) -> bool where F: FnMut(&uint, &T) -> bool {
```
Comment (apasel422) [inline src/trie/map.rs]: Thanks for pointing that out! I've restored the explicit lifetimes and changed `trie_map::test::test_each_reverse` to ensure that your example works.

**Comment by Gankro** [comment]
LGTM, thanks!

**Pull request merged**